modules/FindImGui.cmake

```diff
@@ -0,0 +1,82 @@
+#.rst:
+# Find ImGui
+# -------------
+#
+# Finds the ImGui library. This module defines:
+#
+#  ImGui_FOUND                - True if ImGui library is found
+#  ImGui::ImGui               - ImGui imported target (only on CMake >= 3.0,
+#   because interface targets are not supported on 2.8.12)
+#
+# Additionally these variables are defined for internal usage:
+#
+#  ImGui_INCLUDE_DIR          - Include dir
+#
+
+#
+#   This file is part of Magnum.
+#
+#   Copyright © 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018
+#             Vladimír Vondruš <mosra@centrum.cz>
+#   Copyright © 2018 Jonathan Hale <squareys@googlemail.com>
+#
+#   Permission is hereby granted, free of charge, to any person obtaining a
+#   copy of this software and associated documentation files (the "Software"),
+#   to deal in the Software without restriction, including without limitation
+#   the rights to use, copy, modify, merge, publish, distribute, sublicense,
+#   and/or sell copies of the Software, and to permit persons to whom the
+#   Software is furnished to do so, subject to the following conditions:
+#
+#   The above copyright notice and this permission notice shall be included
+#   in all copies or substantial portions of the Software.
+#
+#   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+#   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#   DEALINGS IN THE SOFTWARE.
+#
+
+list(FIND ImGui_FIND_COMPONENTS "Sources" _INDEX)
+if(_INDEX GREATER -1)
+    set(_FIND_SOURCES TRUE)
+endif()
```

Comment:
Hmm. For a seamless vcpkg compatibility the presence of either the sources or a static lib should be abstracted away from the user. In case of vcpkg it's just `find_package(imgui CONFIG)`, right? Then I would do:

- the "canonical" way to do the above is a "loop switch", i.e. `foreach(_component ${ImGui_FIND_COMPONENTS})`. Then you set `ImGui_Sources_FOUND` accordingly and pass `HANDLE_COMPONENTS` to FPHSA to make it handle the components correctly. Also it's not a problem if a `ImGui::ImGui` or `ImGui::Sources` target is created before the FPHSA call.
- if `find_package(imgui CONFIG)` succeeds, then `ImGui::ImGui` would have just [imgui::imgui](https://github.com/Microsoft/vcpkg/blob/master/ports/imgui/CMakeLists.txt) in its `INTERFACE_LINK_IBRARIES`, plus, if requested, an (otherwise empty) `ImGui::Sources` target, which has `ImGui::ImGui` in its `INTERFACE_LINK_IBRARIES`
- if it fails, then it tries to find the include dir and adds that to `ImGui::ImGui` `INTERFACE_INCLUDE_DIRECTORIES`
- and for sources, if requested, it finds the three `*.cpp` files and creates a new `ImGui::ImGui` target with `INTERFACE_SOURCES` set to these and `ImGui::ImGui` in its `INTERFACE_LINK_LIBRARIES` (and *not* the other way around)

As far as I can see, the only user of `ImGui::Sources` would be `ImGuiIntegration`, which also sets up correct `IMGUI_API` depending on whether it is built dynamically or not. To its `INTERFACE` it would expose only `ImGui::ImGui`, not `ImGui::Sources`.

Hopefully I didn't miss anything blatantly obvious here ... uh, how should it behave regarding dllimport/dllexport if building the integration as dynamic and the vcpkg lib is static? Is that a valid scenario? Maybe in that case add the static lib also to `INTERFACE` link libraries?

Reply:
Yes, that's how I had it implemented until I realised:

What if the user builds with imgui sources and then uses vcpkg for his project? Then there would be duplicate symbols. Same for the other way around.

`IMGUI_API` is already being set up correcty, see [visibility.h](https://github.com/mosra/magnum-integration/pull/33/files#diff-77a8c1458c9c3ca37059a14f80fa00ac).

I guess, we'd put in a big warning of "make sure to link imgui consistently either with vcpkg or IMGUI_DIR (and sources)"

Which all made me decide to not support vcpkg for now, but maybe in a second step.

> the integration as dynamic and the vcpkg lib is static?

The vcpkg lib is dynamic if vcpkg is currently building integration with a dynamic triplet, if it's building static, imgui will be static aswell. Which, yes, means that the symbols will not be present and we would have to somehow merge imgui.lib into MagnumImGuiIntegration.lib using libtool or something. Or would I have the find module still link imgui.lib via `INTERFACE_LINK_LIBRARIES`?

I think assuming that if we found imgui through the package config, MagnumImGuiIntegration will have been build with the package config/precompiled version also should be a more or less safe bet. It's also the most user friendly...

Reply:
> What if the user builds with imgui sources and then uses vcpkg for his project? [...] Same for the other way around.

How common is that? I don't think people are doing this. The other way around I don't see a problem either: it'll find the headers and link to just that, no duplicate symbols.

> we would have to somehow merge imgui.lib into MagnumImGuiIntegration.lib 

No, just have it in `INTERFACE_LINK_LIBRARIES` like I said above.

> I think assuming that if we found imgui through the package config, MagnumImGuiIntegration will have been build with the package config/precompiled version 

Yes. I don't see a value in supporting the very theoretical case of using magnum from outside of vcpkg together with vcpkg pacakges. You'd run into the same problems when combining e.g. PngImporter built outside of vcpkg against some external libPNG and then using that with libPNG from vcpkg.

Reply:
True, alright, will apply changes as suggested.

Reply:
> if find_package(imgui CONFIG) succeeds, then ImGui::ImGui would have just imgui::imgui in its INTERFACE_LINK_IBRARIES, plus, if requested, an (otherwise empty) ImGui::Sources target, which has ImGui::ImGui in its INTERFACE_LINK_IBRARIES

Uhm. But the `ImGui::Sources` would _not_ be in the `INTERFACE_LINK_LIBRARIES`, right? That wouldn't make sense as the symbols would exist twice then, once from the sources and once from the libs in the `imgui::imgui` target.

Reply:
Ah, I see now. The ImGui::Sources is supposed to be empty :)

Reply:
Yes, empty, except for having ImGui::ImGui in its `INTERFACE_LINK_LIBRARIES`.